src/main/scheduler/scheduler.c

```diff
@@ -350,79 +375,143 @@ static void readSchedulerLocals(task_t *selectedTask, uint8_t selectedTaskDynami
 
 FAST_CODE void scheduler(void)
 {
-#if defined(USE_LATE_TASK_STATISTICS)
-    static task_t *lastTask;
-#endif
     // Cache currentTime
     const timeUs_t schedulerStartTimeUs = micros();
-    timeUs_t currentTimeUs = schedulerStartTimeUs;
+    uint32_t nowCycles;
     timeUs_t taskExecutionTimeUs = 0;
     task_t *selectedTask = NULL;
     uint16_t selectedTaskDynamicPriority = 0;
     uint16_t waitingTasks = 0;
-    bool realtimeTaskRan = false;
-    timeDelta_t gyroTaskDelayUs = 0;
+    uint32_t nextTargetCycles = 0;
+    int32_t schedLoopRemainingCycles;
 
     if (gyroEnabled) {
         // Realtime gyro/filtering/PID tasks get complete priority
         task_t *gyroTask = getTask(TASK_GYRO);
-        const timeUs_t gyroExecuteTimeUs = getPeriodCalculationBasis(gyroTask) + gyroTask->desiredPeriodUs;
-        gyroTaskDelayUs = cmpTimeUs(gyroExecuteTimeUs, currentTimeUs);  // time until the next expected gyro sample
-        if (gyroTaskDelayUs <= 0) {
-            taskExecutionTimeUs = schedulerExecuteTask(gyroTask, currentTimeUs);
+        static uint32_t lastTargetCycles = 0;
+        nowCycles = getCycleCounter();
+#if defined(UNIT_TEST)
+        lastTargetCycles = clockMicrosToCycles(gyroTask->lastExecutedAtUs);
+#else
+        if (lastTargetCycles == 0) {
+            lastTargetCycles = nowCycles;
+        }
+#endif
+        if (desiredPeriodCycles == 0) {
+            desiredPeriodCycles = (int32_t)clockMicrosToCycles((uint32_t)gyroTask->desiredPeriodUs);
```

Comment:
when is `desiredPeriodCycles` ever 0?  The `schedulerInit` already has this line:

```
desiredPeriodCycles = (int32_t)clockMicrosToCycles((uint32_t)getTask(TASK_GYRO)->desiredPeriodUs);
```

Reply:
https://github.com/betaflight/betaflight/pull/10573/commits/9be1449cc7d740f3f227afc9a28e032d78aa3830 moves this to `schedulerInit()`